experimental/ast/decl_def.go

```diff
@@ -0,0 +1,541 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ast
+
+import "github.com/bufbuild/protocompile/internal/arena"
+
+// DeclDef is a general Protobuf definition.
+//
+// This [Decl] represents the union of several similar AST nodes, to aid in permissive
+// parsing and precise diagnostics.
+//
+// This node represents messages, enums, services, extend blocks, fields, enum values,
+// oneofs, groups, service methods, and options. It also permits nonsensical syntax, such as a
+// message with a tag number.
+//
+// Generally, you should not need to work with DeclDef directly; instead, use the As* methods
+// to access the correct concrete syntax production a DeclDef represents.
+type DeclDef struct {
+	withContext
+
+	ptr arena.Untyped
+	raw *rawDeclDef
+}
+
+type rawDeclDef struct {
+	ty   rawType // Not present for enum fields.
+	name rawPath
+
+	signature *rawSignature
+
+	equals rawToken
+	value  rawExpr
+
+	options arena.Pointer[rawCompactOptions]
+	body    arena.Pointer[rawDeclScope]
+	semi    rawToken
+}
+
+// DeclDefArgs is arguments for creating a [DeclDef] with [Context.NewDeclDef].
+type DeclDefArgs struct {
+	// If both Keyword and Type are set, Type will be prioritized.
+	Keyword Token
+	Type    Type
+	Name    Path
+
+	// NOTE: the values for the type signature are not provided at
+	// construction time, and should be added by mutating through
+	// DeclDef.Signature.
```

Comment:
Why is this? Seems like it would be easy enough to add them here, since this does require the rest of the declaration has been parsed, all the way through the end of body/semicolon.

Also, `Signature` has no mutator methods for setting the inputs and outputs. Accidental omission? Maybe a TODO is needed?

Reply:
Signature has accessors for Inputs() and Outputs(), which are `Commas[Type]`.